### PR TITLE
GODRIVER-1837 MongoDB 4.2 support for continuous matrix testing

### PR DESCRIPTION
**Pull request by divjotarora**
NOTE: As described in #598, this PR won't actually be merged into the release/1.1 branch. Instead, the changes in this PR will be used to create a new `4.2-continuous-matrix-testing` branch based on release/1.1.

Code changes required:

1. The driver only supported installation via `dep`, which isn't compatible with Go 1.15, so this branch creates a module via `go mod init go.mongodb.org/mongo-driver`. This results in the deletion of `Gopkg.toml`, `Gopkg.lock`, and some files in `vendor` as well as the creation of `go.mod`, `go.sum`, `vendor/modules.txt`.
2. Change in `bsoncore.go` to account for a Go 1.15 change to disallow `string(int)` conversions.
3. Changes in various files to switch single-line functions to multi-line functions due to a change in `gofmt`.

Test changes required:

1. Change in `mongo_test.go` to account for the same `string(int)` conversion invalidation as described in the previous section.
2. Small fix for a test in `default_value_decoders_test.go` due to a Go 1.15 change in the `url` package.
3. JSON unmarshalling change in `server_rtt_test.go` to account for a Go 1.15 change that bans decoding `"null"` into a `json.Number`.
4. Fix a number of tests in `index_view_internal_test.go` because `listIndexes` removed the `ns` field in the index specifications it returns in 4.4.
5. Fix `TestIndexView_CreateOneWithAllOptions` in `index_view_internal_test.go` to only set the `BucketSize` option on versions <= 4.9 because 4.9+ do not support geoHaystack indexes.
6. Fix `TestIndexView_CreateOneWildcard/CreateWildcardIndexWithProjection` in `index_view_internal_test.go` to specify the normalized version of the wildcard projection document due to a change in the format of the documents returned by `listIndexes` in 4.5.x.
7. Same TLS changes for the X509 integration test (`TestClient_X509Auth` in `mongo/client_internal_test.go`) as described in #598.

### CODE REVIEW

File: mongo/index_view_internal_test.go [hunk 1]
```diff
@@ -87,6 +87,25 @@ func getIndexableCollection(t *testing.T) (string, *Collection) {
...
+
+	serverVersion, err := getServerVersion(createTestDatabase(t, nil))
+	require.NoError(t, err)
+	return compareVersions(t, serverVersion, targetVersion)
+}
+
+func assertIndexNSMatch(t *testing.T, expected, actual string) {
+	t.Helper()
+
+	// 4.4+ servers do not return an "ns" field in index specifications.
```
Comment (benjirewis) [inline mongo/index_view_internal_test.go:103]: Perhaps I'm misunderstanding CMT, but do these tests get run against a 4.4+ server? I thought this branch would only be tested against a 4.2 mdb instance.

File: mongo/index_view_internal_test.go [hunk 2]
```diff
@@ -215,32 +234,37 @@ func TestIndexView_CreateOneWithAllOptions(t *testing.T) {
...
+		SetBits(2).
+		SetMax(10).
+		SetMin(1).
+		SetPartialFilterExpression(bsonx.Doc{}).
+		SetStorageEngine(bsonx.Doc{
+			{"wiredTiger", bsonx.Document(bsonx.Doc{
+				{"configString", bsonx.String("block_compressor=zlib")},
+			})},
+		})
+	// 4.9+ servers reject bucket size because they do not support geoHaystack indexes.
```
Comment (benjirewis) [inline mongo/index_view_internal_test.go:258]: Same question here. Does this branch run against `latest`/ `4.9`?

**Review by benjirewis — COMMENTED** [review]
Looks great 😃 ! A few questions, though.

Also, what are the new empty shell script files added under `vendor`? Looking at `go-toml/benchmark.sh`, `go-toml/fuzz.sh`, `x/sys/unix/mkerrors.sh`, etc. Just pieces of dependency code that are empty?

File: mongo/index_view_internal_test.go (hunk 1 above)
Comment (divjotarora) [reply mongo/index_view_internal_test.go:103]: The 4.2 driver was already tested against the 4.2 server when we were developing it. The goal of this project is to test "old" drivers against "new" servers to ensure that customer applications running with an older driver version like this one will continue to function if the server is upgraded. To do this, we test driver version X against all server versions > X. For example, 4.0 drivers get tested against 4.2, 4.4, and latest.

File: mongo/index_view_internal_test.go (hunk 2 above)
Comment (divjotarora) [reply mongo/index_view_internal_test.go:258]: See the response above.

File: mongo/index_view_internal_test.go (hunk 1 above)
Comment (benjirewis) [reply mongo/index_view_internal_test.go:103]: Ah that makes sense; the 4.2 branch just represents the state of the driver as of server version 4.2. Thanks for the explanation.

**Review by benjirewis — APPROVED** [review]
LGTM 🎉

**Review by iwysiu — APPROVED** [review]

**Comment by divjotarora** [comment]
I'm going to close this PR and push up a 4.2-continuous-matrix-testing branch to `origin`.

**Pull request closed without merging**